### ISSUE TITLE
Integrate clippy checking in CI build

### ISSUE THREAD

**Issue by abrown** [CraneStation/cranelift#1170]
Please try to describe precisely what you would like to do in Cranelift and/or
expect from it. You can answer the questions below if they're relevant and
delete this text before submitting. Thanks for opening an issue!

- What is the feature or code improvement you would like to do in Cranelift?
After merging #1169 (and seeing #1168), it occurred to me that we should output clippy warnings as a part of `test-all.sh`. I'm not sure we should fail the build (though this could be an option) but this would make non-clippy-approved additions more visible.

- What is the value of adding this in Cranelift?
It should result in cleaner, more standard code in the long run.

- Do you have an implementation plan, and/or ideas for data structures or
  algorithms to use?
Add `cargo clippy` to `test-all.sh`.

- Have you considered alternative implementations? If so, how are they better
  or worse than your proposal?
No.

**Comment by abrown** [comment]
@bnjbvr, this seems like a relatively no-brainer to me but I guess the question is: do we fail the build on clippy warnings or not?

**Comment by fitzgen** [comment]
In general, I feel like it is good to fix clippy warnings (or ignore them in specific cases) but rarely useful to hard gate merging PRs on being clippy clean.

**Comment by peterhuene** [comment]
I agree that it shouldn't be required, but perhaps an optional CI leg that may fail and bring some visibility to the accumulation of the warnings?  It looks like it's been some time since they were last addressed and I think it's important for upstream crates to be free from linter warnings so that downstream crates don't get inundated with unfixable warnings.

**Comment by bnjbvr** [comment]
Agreed it's a nice idea in general and that it shouldn't block PR from being merged, though. We could add an optional job in the CI, that would have a `continueOnError` directive, as we've recently done for the nightly fuzzing job.

That being said, I question just a bit the utility of such CI jobs: there's no real incentive to fix them, apart from wanting to maintain great quality. It's a nice source of easily-fixable bugs, though.

Another approach would be to remove the `clippy.toml` file that's at the top-level: most IDEs see it and enable clippy warnings by default in the IDE. Removing the file would hide IDE warnings, and thus remove the cognitive load associated to them.

My personal preference goes to a `continueOnError` CI job right now.

**Comment by abrown** [comment]
Just looked at this a bit more and I think someone with rights should install https://github.com/marketplace/actions/rust-clippy-check. That GitHub action will use clippy to check PRs and post any issues as annotations (follow the previous link for an example). I think to use this we need to use a GitHub secret token, see https://github.com/marketplace/actions/rust-clippy-check#example-workflow.

**Comment by bnjbvr** [comment]
Interesting. I'm only wondering what this will do to existing Clippy warnings: will they be reported as part of a PR that modifies code around, or will only new Clippy warnings be reported? In the former case, I think it would be nicer to first fix all the Clippy warnings before enabling such a check.

**Comment by abrown** [comment]
I suspect we would have to fix all the clippy warnings first but that wasn't too difficult: #1340.

**Comment by abrown** [comment]
@alexcrichton, what do you think about adding this to cranelift and wasmtime, perhaps when they merge?

**Comment by alexcrichton** [comment]
I personally agree with @fitzgen's [thoughts above](https://github.com/bytecodealliance/cranelift/issues/1170#issuecomment-546428127) where hard-gating on clippy on CI is often quite onerous. I personally disagree with a lot of clippy's lints as well which doesn't help my own opinion though... I would advocate for a "here's how to run it if you want" at most, but otherwise don't hard-gate it on CI.

**Comment by abrown** [comment]
👍 I agree with the "no hard-gating" point of view as well. I think the point of installing this clippy action would be for awareness more than anything else: instead of running clippy in CI, ignoring failures, and thus doing extra work that is ignored, this action would make PR submitters and reviewers aware of what clippy thinks is problematic. If clippy finds a real issue--good; if clippy finds irrelevant stuff then they can ignore the warning (locally or globally).

**Comment by philipc** [comment]
The limitations on https://github.com/marketplace/actions/rust-clippy-check make it seem to be not worth the effort: "Due to token permissions, this Action WILL NOT be able to post clippy annotations for Pull Requests from the forked repositories."

**Comment by abrown** [comment]
Hm, yeah, then there's not much point to doing this until that is fixed; seems like this can issue can go forward when/if https://github.com/actions-rs/clippy-check/issues/2 is fixed.
